### PR TITLE
add yoloSegment head

### PR DESCRIPTION
**Pull request by AyushExel**
@laughing-q I wanted to experiment with this new head but there's something causing cuda problems. It works fine on my non gpu system but fails on the server with cuda enabled gpus. I think its something to do with amp checks. 
The amp checks fail and this error pops up.
Screenshot 2022-07-29 at 9 55 49 AM
Screenshot 2022-07-29 at 9 56 13 AM

Do you have any ideas? Maybe I'm using some operation not compatible with amp, but I'm not an expert in using mixed precision

### CODE REVIEW

**Comment by Laughing-q** [comment]
@AyushExel this will work. 
```python
        self.mask_conv = nn.ModuleList([
            # layer 0 - for anchor 1
            nn.Sequential(
            nn.Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1)),
            nn.Upsample(scale_factor=2.0, mode="bilinear",align_corners=False),
            nn.Conv2d(256, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1)),
            nn.SiLU(inplace=True),
            ),

            # layer 1 - for anchor 2
            nn.Sequential(
            nn.Upsample(scale_factor=2.0, mode="bilinear", align_corners=False),
            nn.Conv2d(256, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1)),
            nn.Upsample(scale_factor=2.0, mode="bilinear", align_corners=False),
            nn.SiLU(inplace=True)),

            # layer 2 - for anchor 3
            nn.Sequential(
            nn.Upsample(scale_factor=2.0, mode="bilinear", align_corners=False),
            nn.Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1)),
            nn.Upsample(scale_factor=4.0, mode="bilinear", align_corners=False),
            nn.SiLU(inplace=True)),


            # mask layer
            nn.Sequential(
            nn.Conv2d(512, mask_dim, kernel_size=1, padding=0),
            nn.SiLU(inplace=True))])

```

**Comment by AyushExel** [comment]
@Laughing-q thanks. Didn't know that nn.sequential couldn't be used independently. 
Anyway, the size of the model went up by a lot.. yolov5s went from 7M params to 12M.. If this performs well, we'll still need to reducde the convolution layers, or use some tricks to reduce params. I think one convolution per anchor should also be enough which takes nxwxh --> 512x64x64, so all of them can be added

**Comment by AyushExel** [comment]
@Laughing-q here's the workspace for the runs with yolosegment head https://wandb.ai/cayush/yolov5-yoloseg/

**Comment by Laughing-q** [comment]
> @Laughing-q thanks. Didn't know that nn.sequential couldn't be used independently. Anyway, the size of the model went up by a lot.. yolov5s went from 7M params to 12M.. If this performs well, we'll still need to reducde the convolution layers, or use some tricks to reduce params. I think one convolution per anchor should also be enough which takes nxwxh --> 512x64x64, so all of them can be added

yes, I think so

**Comment by AyushExel** [comment]
@Laughing-q I'm iterating on the design of the mask head.. Nothing impressive yet but I'll keep experimenting

**Comment by AyushExel** [comment]
@Laughing-q I tried a bunch of heads but nothing came close to performing even as good as the original. I'm kind of out of ideas.. solov2 wasn't a great help. Maybe we should build something based on some other architecture

**Comment by AyushExel** [comment]
I think the main point of improvement that we can make is somehow incorporate the information of 2nd and 3rd anchors in the conv_mask.. This is what I tried to do with convoluting them to 64x64 and adding them.. But maybe they are too small for that complex operation and all information is lost in the process.
EDIT: Now that I think about it, maybe the anchors mostly contain information about the boxes and not the segments..maybe we can branch out the segment head from some previous layer that contains more information about the features?

**Comment by Laughing-q** [comment]
@AyushExel As for the incorporating information of 2nd and 3rd anchors, the neck from yolov5 is PANet right? 
 actually the original head has integrated information from other layers(the first top-down path). and maybe the head you updated was making `proto_out` a little complicated which the mask coefficient(mask_dim=32) couldn't handle it, after all, solov2 has CoordConv to deal with its incorporated feature, just a guess.

**Comment by AyushExel** [comment]
Oh yes. I'm using the wrong terminology. I meant the other layers in from the input of detect layer. Btw I ran the run with the original head and its performing much better. I'll keep experimenting for a day or two..
I was thinking about updating the code to pytorch 1.12.1 because that supports deterministic behaviour and we can be sure if something is better or worse without testing it multiple times

**Comment by Laughing-q** [comment]
> I think the main point of improvement that we can make is somehow incorporate the information of 2nd and 3rd anchors in the conv_mask.. This is what I tried to do with convoluting them to 64x64 and adding them.. But maybe they are too small for that complex operation and all information is lost in the process. EDIT: Now that I think about it, maybe the anchors mostly contain information about the boxes and not the segments..maybe we can branch out the segment head from some previous layer that contains more information about the features?

Ooh sorry, I didn't see the `EDIT` part, and I think this is a idea.

**Comment by Laughing-q** [comment]
@AyushExel and I saw you changed `mask_dim` to 64 just now, there might be a bug, you have to update this in `evalutor.py`
[https://github.com/AyushExel/yolov5/blob/a7e58685f2a52187c5b7fb159df008b90a09548e/evaluator.py#L306](https://github.com/AyushExel/yolov5/blob/a7e58685f2a52187c5b7fb159df008b90a09548e/evaluator.py#L306)
```python
        out = self.nms(prediction=out, conf_thres=self.conf_thres, iou_thres=self.iou_thres, multi_label=True,
            agnostic=self.single_cls, mask_dim=de_parallel(model).model[-1].mask_dim)
```
and function `de_parallel` is in `utils/torch_utils.py`

**Comment by AyushExel** [comment]
@Laughing-q ok thanks.. I haven't started a run for that yet. I'll update as you said.. I'm just trying to see what happens on increasing the size..

**Comment by AyushExel** [comment]
@Laughing-q I see that there are a few layers commented out of the the DetectSegment head. Did you try those layers and they caused no improvement? Or should I try a run with those ?

**Comment by Laughing-q** [comment]
@AyushExel I haven't tried that yet, you can test it if you want.

**Comment by AyushExel** [comment]
Okay that didn't do anything.. Did you make any progress in finding the batch-size bug?

**Comment by Laughing-q** [comment]
@AyushExel not yet... :( 
I'll continue trying to fix it today...

**Comment by AyushExel** [comment]
@Laughing-q same here.. No progress on the head.

**Comment by AyushExel** [comment]
@Laughing-q okay no luck on the head.
The main reason why I wanted to experiment with head was to investigate why yolov5l was performing worse than yolov5m..Also, the fact that you pointed out that the map difference in box and mask seems too high in these models as compared to other models.. 
I don't really know what could be the cause now that I've experimented a bit with new and existing seg head. So, maybe we should just launch yolov5s and m models for segmentation. What do you think ?

**Comment by Laughing-q** [comment]
@AyushExel I got the same situation with batch-size :(
of course we can just launch yolov5s and yolov5m if you think the accurary is good. But I think we should still try to improve the accuracy, also the bs issue...

**Comment by AyushExel** [comment]
@Laughing-q oh yes, maybe the bs issue is more important.. so maybe we'll need to fix that for launch.. But regarding accuracy.. I think on paperswithcode, realtime instance segmentation sota is 35mAP and yolov5m model has 32-33mAP right? So it should be fine.. But we do have all of august to make it better.. the plan is to finalize this by the end of august

**Comment by Laughing-q** [comment]
@AyushExel yes, and actually the yolov5m-seg model is much faster than others.

**Comment by AyushExel** [comment]
@Laughing-q yeah thats great! I was so excited about the l model because of that. If m model is so close to the sota and much faster, I thought L would go beyond the sota and still be much faster 😂

**Comment by AyushExel** [comment]
@Laughing-q Ok I think I found something interesting.
* Uncommenting the extra layers cause a significant performance improvement in m models.
* Removing overlap causes the models to perform much worse.. This is surprizing, maybe related to bs bug?
[Link](https://wandb.ai/cayush/yolov5-yoloseg) to dashboard.. Here, the runs with suffix `feat64` are models with extra layers, see how they perform much better in first 50 epochs( I only trained for 50), maybe they'll perform better with more epochs.. And same model on removing overlap performs much worse.

**Comment by Laughing-q** [comment]
@AyushExel okay thanks, I'll check overlap option too.

**Comment by AyushExel** [comment]
@Laughing-q I've run multiple runs and I think I can now verify that the new head is providing a solid 2% increase in mAP for m models.
But note that I'm training from pretrained yolov5 checkpoints for only 50 epochs, so its possible that it performs much better.

**Comment by AyushExel** [comment]
@Laughing-q Have you reached any conclusion for the batch-size bug? 
I was thinking of running a final benchmark with the updated head for s and m model. But before that I want to make sure we've tried everything we can think of.

And also, there are 2 additional things I think we should do before releasing:
* Overlap masks for pred also
* alignment with cocoapi scores. There's too much difference b/w our and cocoapi score. Think you said that cocoapi does evaluation in full resolution thats why it has higher score but I don't remember clearly. We should either perform final evaluation using cocoapi or improve our eval to get close to cocoapi.. Or both..  What do you think?

I haven't looked at the overlap-mask code so probably you'll need to help me with that.. But I can take on the cocoapi integration :)

**Comment by Laughing-q** [comment]
@AyushExel Hey, finally I find the batch-size issue, actually there is no bug,  I'm pretty sure it's an optimization problem, and this took me so long.....
I just check everything in the code and couldn't find anything wrong, and I used a large batch-size today and got good results accidentally, I found that it was random, but the probability of getting bad results was pretty higher, then I found the loss did not perform well in the first few epochs. So I shut the warmup down and reduced basic learning rate, then the model preformed well with large batch-size.
So I think we should find a new lr for large batch-size or test a new lr for current batch-size before we build the benchmark, the current lr may not be the best for segmentation. what do you think?
And also the overlap mask for predictions, yes I think so, I'll do that later, maybe after we find a new optimization strategy.

**Comment by AyushExel** [comment]
@Laughing-q Ohh wow.. I think I expected that we'll need to find better hyperparameters but I also thought that most of the network is same as detection so hyperparameters should not have a major effect.. Now that we know that its a hyper-parameter problem, I'll stop the training and I'll try to optimize the hyper-parameters.. I'll think of a strategy as I don't have a plan for that right now , but I think evolving will be too slow for such large network and dataset.. Maybe some other hyper-parameter optimization strategy? I'll research a bit

**Comment by Laughing-q** [comment]
@AyushExel I also think that finding the best value for the whole hyper-parameters will take so long, maybe we can start with learning rate or something, like, we find a basic lr for a basic batch-size, then scale the lr according to the batch-size, like the mmdetection [https://github.com/open-mmlab/mmdetection/blob/3b72b12fe9b14de906d1363982b9fba05e7d47c1/mmdet/apis/train.py#L107](https://github.com/open-mmlab/mmdetection/blob/3b72b12fe9b14de906d1363982b9fba05e7d47c1/mmdet/apis/train.py#L107)

**Comment by AyushExel** [comment]
@Laughing-q ok that also sounds great! Btw, I just had a thought that maybe L model is suffering because the same issue?

**Comment by Laughing-q** [comment]
@AyushExel yes, it could be.

**Comment by AyushExel** [comment]
@Laughing-q ok, I'll Try training with a higher LR and check

**Comment by Laughing-q** [comment]
@AyushExel Some good news here, I did some experiments and you can see it here [https://wandb.ai/laughing/test](https://wandb.ai/laughing/test), the metrics are so low which you can ignore, just check the predicted masks.
now I'm pretty sure it's about one hyper-parameter `warmup_bias_lr`, this hyp is too high as 0.1. Maybe it's suit for detection, but not for segmentation.

**Comment by AyushExel** [comment]
@Laughing-q ok great! I'll try with the reduced lr value in the L model. Let's see if that does something. Other than that I think we can just scale down the warmup-lr with bs.

**Comment by AyushExel** [comment]
@Laughing-q Ok so I ran a run with all hyps-high.yaml only changed warmup to 0.01 and the L model is performing as good as the original M model as this point. I'll wait for all 50 epochs.. The M models with extra uncommented layers still perform the best but the performance difference in L model with and without reduced warmup is huge.. after this experiment, I'm pretty sure the L model can perform better than m if we just find the right hyps.

**Comment by Laughing-q** [comment]
@AyushExel That's great! And also I found reducing `warm_bias_lr` is not friendly to yolov5s-seg with a small batch-size. And about the overlap thing, here's what I'm thinking:
the non-overlap predicted masks might slow down in three ways: 
- calculation of loss function, actually I think the calculation of loss is fast enough now, and it's a hard change with the code if we make `pred_masks` overlap and this might harm the accuracy.
- plotting, as we just plot the top 10 pred_masks in the first few batches, the speed is acceptable.
- saving coco'json for evaluation, actually this could get really slow, but the overlap `pred_masks` might be harmed, but evaluation need the good predictions, so there may be no way to avoid it.
Also if we have to do something about the overlap predicted masks, maybe just the plotting way, cause it's just a preview of masks, which doesn't need the accuracy. what do you think?

**Comment by AyushExel** [comment]
@Laughing-q Yeah I agree. Overlapping pred masks only for plotting sounds great.
I did not quite get your point about evaluation using json. Do you think we should not do that?

**Comment by Laughing-q** [comment]
@AyushExel Ooh, we definitely should do that, I mean that we can't avoid the slow speed when we do evaluation with cocoapi.

**Comment by AyushExel** [comment]
@Laughing-q hmm yeah.. I haven't used it yet, the server is always busy with some model training. How much time does it take usually? Also, it should be fine if we just use it for final evaluation?

**Comment by Laughing-q** [comment]
@AyushExel It's about 12 minutes, batch-size=8 on my RTX3060 notebook.

**Comment by AyushExel** [comment]
@Laughing-q okay that's too slow.. I think that's why we don't use cocoapi for detection evaluation. But those results are very close to cocoapi already. Segmentation is a bit different though.

**Comment by AyushExel** [comment]
@Laughing-q so I trained 2 L models.. One with reduced warmup_lr and one with both reduced warmup_lr and Adam optimizer with lr 0.001.. Both performed similar to the M model. I'm still trying to get the L model to perform better than the M model..

**Comment by Laughing-q** [comment]
@AyushExel Hey, good news, I got a significant improvement with yolov5s-seg, also reduce the gap of mAP between bbox and masks. Check this [link](https://wandb.ai/laughing/runs?workspace=user-laughing) and check the `yolov5s-mr4-overlap-tobj-0.5-newlabels` and `yolov5s-mr4-overlap-base`. And I need a extra experiment to confirm this, so if this work, I'll make a new PR to update this(maybe 1 or 2 hours later), and this might also make yolov5l-seg better, what do you think?

**Comment by AyushExel** [comment]
@Laughing-q that's such a huge improvement.. Looks like even s model will outperform sota when evaluated using cocoapi.
On my end I was able to push L model's performance beyond m model finally.
[Link](https://wandb.ai/cayush/yolov5-yoloseg?workspace=user-cayush).. the red model is the best even performing m model..
I tried these things with l model.
* first only reduced warmup - improved performance
* Reduced warmup and increased lr - broke the model
* Tried adam with reduced warmup - didn't work
* Tried adam with reduced warmup and reduced lr - slightly worse performace than m model
* Tried adam with original warmup and reduced lr (latest run)- Started to perform better than all models but the curve finally settled to perform slightly worse than the best m model.

Now maybe this combined with your experiments can cause major improvement in L model also

**Comment by AyushExel** [comment]
@Laughing-q hey, can you take a look the latest dashboard [link](https://wandb.ai/cayush/yolov5-seg-newlabels?workspace=user-cayush) that I shared for the training of m model from scratch.
it seems like it's going to perform similar to my [previous m benchmark](https://wandb.ai/cayush/yolov5-seg?workspace=user-cayush).. 
Are you overriding some hyps in your training or are you using all the default ones?

**Comment by Laughing-q** [comment]
@AyushExel I think the latest training is better, you can compare the `seg_loss` and the mAP gap between bbox and masks. The latest one has lower `seg_loss` and smaller mAP gap. So I think the accuracy will be better later.

**Comment by Laughing-q** [comment]
@AyushExel Ooh, and I noticed the latest training is mr4 and the older training in your dashboard is mr1, it's normal that mr1 shows higher mAP compared with mr4 in the dashboard(cause the mask size). So as the mAP is almost the same now, the mr4 definitely outperform the mr1, maybe you can check it with `val_instseg.py`.
Also there's one unusual thing, the bbox mAP seems a little lower than before, and I didn't find the same thing with s model.

**Comment by AyushExel** [comment]
@Laughing-q ohh yeah!! you're right about the mr.. It just slipped out of my mind. But regarding the low map for boxes - I think one stop solution for all our reproducibility problems is the switch to torch 1.12.1 because then yolov5 automatically switched to deterministic behaviour.
https://github.com/ultralytics/yolov5/blob/4d8d84b0ea7147aca64e7c38ce1bdb5fbb9c5a53/utils/general.py#L203

**Comment by Laughing-q** [comment]
@AyushExel okay, I'll switch to torch 1.12.1 after I finish yolov5s-seg training. Also if you find the bbox mAP is too low to accept, maybe revert `seg_loss.py` will help. The update in `seg_loss.py` get a little higher mask mAP and the same bbox mAP with s model, but I'm not so sure the m model will behave the same.
And all the loss is lower then old training except bbox, this is strange. I think there's something wrong with the bbox.
